libgadget/fof.c

```diff
@@ -382,40 +373,96 @@ void fof_label_primary(void)
     myfree(FOF_PRIMARY_GET_PRIV(tw)->Head);
 }
 
+/* This function does the actual merging of the trees. It takes two head particles, h1 and h2,
+ * and merges 2 into h1 by setting Head[h2] = h1. It does this if and only if h1 < h2 (to avoid cycles)
+ * and it does it under a lock for h2, so that h2 cannot be merged elsewhere. h1 is not locked, to avoid deadlocking.
+ * It can thus cease to be the head of the tree, but this is fine, because it will never be merged into h2.
+ * If h1 > h2, the code swaps them and tries again.
+ * returns: 1 if a merge happened, 0 if it did not. No locks are held after return.*/
+static int
+try_merge_trees(int * h1, int * h2, MyIDType * subminID, MyIDType * subminIDTask, int * Head)
+{
+    /* We are about to change the HEAD of h2, so lock it.
+     * We need to make sure that h2 is really the head of the subtree,
+     * or not all particles will be linked.
+     * If this is not the case, unlock the particle, walk the tree again,
+     * and loop until we can get a lock.*/
+    while(1) {
+        /*Stop if we encountered h1 along the head of this tree: means this is already merged.*/
+        if(*h2 == -1)
+            return 0;
+        lock_particle(*h2);
+        if(Head[*h2] != *h2) {
+            unlock_particle(*h2);
+            *h2 = HEADa(*h1, *h2, Head);
+        }
+        else
+            break;
+    }
+    /*Get the current subtree properties: this is fine because we are under the lock.*/
+    *subminID = HaloLabel[*h2].MinID;
+    *subminIDTask = HaloLabel[*h2].MinIDTask;
+
+    /*Merge the subtree*/
+    if(*h2 > *h1) {
+        #pragma omp atomic write
+        Head[*h2] = *h1;
+        unlock_particle(*h2);
+        return 1;
+    }
+    else {
+        /*If h1 > h2, swap the particles and try again*/
+        unlock_particle(*h2);
+        int t = *h2;
+        *h2 = *h1;
+        *h1 = t;
+        return try_merge_trees(h1, h2, subminID, subminIDTask, Head);
```

Comment:
Would it not work if passing in h2, h1 instead of h1, h2?

Also is it important this try is after ensuring h2 is the precise head? Given that you have unlocked `*h2`, it does not seem to be important. But then why is the section after the h2 promotion code.

Reply:
Passing in h2, h1 wouldn't swap the heads in the calling function, so MinID wouldn't be set correctly.

I unlock h2 because once we swap them it is the former h1 that needs to be promoted and locked. I think that if I swap first and them promote, the promoted h2 may again be less than h1 (because the indices decrease as we go up the tree). I might be wrong.